docs/src/proposals/blockstore-rocksdb-compaction.md

```diff
@@ -0,0 +1,234 @@
+there're several issues in the Solana's BlockStore which runs RocksDB with
+level compaction.  Here's a quick summary of the issues:
+
+### Long Write Stalls
+Remember that RocksDB periodically runs background compactions in order to
+keep the number of logical levels small in order to reach the target read
+amplification.  However, when the background compactions cannot catch up
+the write rate, the number of logical levels will eventually exceeds the
+configured limit.  In such case, RocksDB will rate-limit / stop all writes
+when it reaches soft / hard threshold.
+
+In [#14586](https://github.com/solana-labs/solana/issues/14586), it is reported
+that the write stalls in Solana's use case can be 40 minutes long.  It is also
+reported in [#16234](https://github.com/solana-labs/solana/issues/16234) that
+writes are also slowed-down, indicating the underlying RocksDB instance has
+reach the soft limit for write stall.
+
+### Delections are not Processed in Time
+Deletions are processed in the same way as other write operations in RocksDB,
+where multiple entries are merged / deleted during the background compaction.
+Although the deleted entries will not be visible from the read side right
+after the deletion is issued, the deleted entries (including the original
+data entries and its deletion entries) will still occupy disk storage.
+
+TBD: explain how write-key order makes this worse
+
+### High Write I/O from Write Amplification
+In addition to write stalls, it is also observed that compactions cause
+unwanted high write I/O.  With the current design where level compaction
+is configured for BlockStore, it has ~30x write amplification (10x write amp
+per level and assuming three levels in average).
+
+## Current Design
+In the current BlockStore, The default level compaction + slot-ID based
+compaction filter + periodic compaction are used, which has ~30x write
+amplification (10x write amp per level and assuming three levels in
+average + the write amp caused by the periodic compaction.)
+
+## Proposed Design -- FIFO Compaction with TTL + Rate Limiting
+As all the above issues are compaction related, it can be solved with
+a proper compaction style and deletion settings.  Fortunately, Solana's
+BlockStore has an unique workload where write-keys are mostly monotonically
+increasing over time.  This allows data to be persisted in sorted order
+naturally without compaction
+
+In the proposed design, we will laverage such unique property to aggressively
+config RocksDB to run as few compactions as possible while offering low read
+amplification with no write stalls.
+
+### FIFO Compaction
+FIFO Compaction is one of the three compaction styles offered by RocksDB which
+actually does not run any compaction in background.  Instead, it simply throw
+aways files from old to new when one of the following conditions meets:
+
+- If TTL is set for the FIFO Compaction, RocksDB will throw away files which is
+  older than the configured TTL.
```

Comment:
We could extend the TTL to be long enough to hold the older ledger.  Btw, can I know more about what is ledger-tool?

Reply:
Yep, ledger-tool is a CLI tool that allows manual inspection / manipulation of a ledger. It is in the monorepo at https://github.com/solana-labs/solana/tree/master/ledger-tool, just run `cargo run -- --help` in that sub-folder and you can see the types of commands that it supports.

There are a bunch of read-only operations as well as a few write operations. Obviously the write operations can modify state, but even these write operations are very targeted (ie delete a single slot). In general, I think it be good to have compactions disabled when using the tool.

One caveat that I'll mention is that you can use ledger-tool on a ledger that is actively being updated by a running validator process. In this instance, I think the RocksDB session from the validator would have compaction enabled, and I think this is ok. But, if I just have a ledger on disk sitting around for a week, I wouldn't want simply opening it to print some statistics to cause data to get dropped because it has surpassed a TTL.

Reply:
https://github.com/solana-labs/solana/issues/19952 captures the sentiment I was getting at, and apparently our current implementation has shortcomings

Reply:
> #19952 captures the sentiment I was getting at, and apparently our current implementation has shortcomings
For the ledger tool, we should open the DB with read-only mode.